posts/2020-10-12-test-graalvm.md

```diff
@@ -0,0 +1,167 @@
+
+[Gitlab CI](https://docs.gitlab.com/ee/ci/) is part of [Gitlab](https://about.gitlab.com/) although you don't need to use the
+latter to use the CI. I already knew it because I used it a lot in my previous job and I really love it.
+
+The idea was to create a CI pipeline with the following objectives in mind:
+
+- Execute it every periodically vs every commit.
+- Only if there were changes in either Micronaut or GraalVM.
+- Compile GraalVM from `master` branch, so we could detect any issue or regression when they were introduced and not in the next release.
+- Test different Micronaut applications easily.
+- Different CI stages and jobs with some dependencies between them.
+
+After a week of work in [this new repository](https://gitlab.com/micronaut-projects/micronaut-graal-tests) and after tried different
+approaches, this was the first [CI pipeline](https://gitlab.com/micronaut-projects/micronaut-graal-tests/-/pipelines/43802345)
+ that passed the build:
+
+![First CI pipeline](2020-10-12-first-ci-pipeline.png)
+
+It has four different stages that we still use:
+
+- **Log-commits**: Simply save the commits that triggered the build for both Micronaut and GraalVM. If the build fails we can
+check those commits and see why it failed.
+- **Build-graal**: Clone the GraalVM repository and build it from `master` branch. The GraalVM JDK distribution generated in
+this stage is saved as an _artifact_ to use it in the next stage. 
+- **Micronaut**: Clone different Micronaut test applications and build a native image using the GraalVM distribution created 
+in the previous stage. The native images are saved again as artifacts, so they can be used for the tests.
+- **Test**: Start the native images generated in the previous stage and run some tests to make sure everything works as expected.
+These tests are `curl` requests to different endpoints and we check the response.
+
+
+For keeping track of new commits in Micronaut and GraalVM, and running the scheduled jobs, we have
+[another companion project](https://gitlab.com/micronaut-projects/micronaut-graal-tests-scheduler) in which we store the
+ latest commit for both projects and decide if we trigger the CI in the main repository.
+
+
+## Test applications
+
+Over this time a lot of things have changed, and those initial two applications have become 26 test applications with 39
+different jobs. This is because some of them test the same thing but for different options. For example in the 
+[Micronaut JDBC](https://github.com/micronaut-graal-tests/micronaut-data-jdbc-graal) test application, there are branches
+for the all the databases supported: H2, Postgres, Oracle, MariaDB, SQL Server and MySQL.
+
+Also, the number of jobs have been multiplied by four because we now test GraalVM for JDK 8 and JDK11, and we also test
+ at two Micronaut versions (at this moment `2.0.x` and `2.1.x`). This is how the CI pipeline looks like today:
+
+
+![CI pipeline today](2020-10-12-ci-pipeline-today.png)
```

Comment:
This image is really large.   Is there a way we can get this across without such a large image?

Reply:
Yeah, I wasn't sure about it either. I've pushed a new version moving the JDK11 tests to a new column on the right. Now everything fits in the screen without doing any scroll:

![image](https://user-images.githubusercontent.com/559192/95558252-3e8e3f00-0a16-11eb-9dd7-12a38ea4c47e.png)